File: AlgorithmicsSet/~PREOI Finał 2019/Dzień III/Teleportacja kwantowa/c.cpp
#include <bits/stdc++.h>

using namespace std;

template<typename T>
using min_queue = priority_queue<T, vector<T>, greater<T>>;

uint32_t adiff(uint32_t a, uint32_t b) { return a > b ? a - b : b - a; }

int main()
{
    uint32_t H, W, n;
    cin >> H >> W >> n;

    vector<uint32_t> Y(n), X(n);
    for(uint32_t i = 0; i < n; i++)
        cin >> Y[i] >> X[i];

    vector<bool> done(n);
    vector<uint32_t> opt(n, -1u);
    vector<uint32_t> track(n, -1u);

    track[0] = -1u; opt[0] = 0;

    min_queue<pair<uint32_t, uint32_t>> Q;
    Q.push({0, 0});

    while(not Q.empty())
    {
        uint32_t u = Q.top().second; Q.pop();
        if(done[u]) continue;
        done[u] = true;
        //cerr << "@ " << u+1 <<  " (" << opt[u] << ")" << endl;
        for(uint32_t v = 0; v < n; v++) if(not done[v])
        {
            uint32_t w = max(adiff(X[u], X[v]), adiff(Y[u], Y[v]));
            if(opt[u] + (1u << w) < opt[v])
            {
                opt[v] = opt[u] + (1u << w);
                track[v] = u;
                //cerr << "push " << v+1 << " with priority " << w << endl;
                Q.push({opt[v], v});
            }
        }
    }

    stack<uint32_t> result;
    uint32_t u = n-1;
    while(u != -1u)
    {
        result.push(u);
        u = track[u];
    }
    cout << result.size() << "\n";
    while(not result.empty())
        cout << result.top()+1 << " ", result.pop();
}

File: AlgorithmicsSet/~PREOI Finał 2019/Dzień III/Teleportacja kwantowa/c.py
import heapq

H, W, n = [int(x) for x in input().split()]

X, Y = [0]*n, [0]*n

for i in range(n):
    p = input().split()
    X[i] = int(p[0])
    Y[i] = int(p[1])

done = [False for _ in range(n)]
opt = [4 << (max(W, H)) for _ in range(n)]
track =  [-1 for _ in range(n)]

opt[0] = 0

heap = []
heapq.heappush(heap, (0, 0))

while heap:
    u = heap[0][1]
    heapq.heappop(heap)
    if done[u]: continue
    done[u] = True
    for v in range(n):
        if not done[v]:
            w = max(abs(X[u] - X[v]), abs(Y[u] - Y[v]))
            if opt[u] + (1 << w) < opt[v]:
                opt[v] = opt[u] + (1 << w)
                track[v] = u
                heapq.heappush(heap, (opt[v], v))

result = []
u = n - 1
while u != -1:
    result.append(u)
    u = track[u]

print(len(result))
print(" ".join(str(x+1) for x in reversed(result)))
